### ISSUE TITLE
Allow use of custom execute flags in objects that don't depend on module where they are defined

### ISSUE THREAD

**Issue by bwspenc** [idaholab/moose#10650]
## Rationale
If a custom exec flag is registered in a module or application, it can only be used within objects that depend on that module or application.

## Description
Arbitrary objects, such as framework objects, should be able to be executed at custom times, not just those defined in the module that defines the custom exec flag.

## Impact
Currently, to enable an object to be executed at custom times, one must add available flags by calling the`ExecFlagEnum::addAvailableFlags()` method within the code. This requires a code dependency on the code that defines the flag, which makes it so that objects that don't depend on the module where that flag is defined can't be executed. Aside from that, it would be very cumbersome to have to do this for every object that could conceivably be executed at a custom time. The architecture of this system should be modified to not require that.

**Comment by bwspenc** [comment]
@permcody @aeslaughter @friedmud

**Comment by aeslaughter** [comment]
This requires two steps: 

1. Change MooseUtils::getDefaultExecFlagEnum to create an ExecFlagEnum with new values allowed.
2. Add a check that will make sure that the value in the MooseObject enum exists in the App enum.

**Comment by permcody** [comment]
One design might be to promote the `MooseUtils::getDefaultExecFlagEnum()` to `MooseApp` (but not make it virtual). You should be able to figure out which applications have been registered and query the flags from all of them. Think about the combined module for instance. This method should be able to loop over all of the "apps" which would consist of MOOSE and all the other modules add append their flags together. This is more complex than just adding a new virtual method which wouldn't do this.

**Comment by aeslaughter** [comment]
This opens up the reason I did not do this, if you just use all registered flags in the object then it is possible to set the flag of an object to something that doesn't ever execute. For example, the EXEC_FAILED flag is only used by a Output objects.

**Comment by aeslaughter** [comment]
Do we need an alternative way to set "execute_on" that allows us to set the available flags like we are now so that we don't have non-executed options?

Perhaps a "custom_execute_on" that can be set to anything. Maybe it also requires "execute_on" to be set to "NONE". Just throwing out ideas.

**Comment by bwspenc** [comment]
@aeslaughter Is it really a problem if you set an exec flag of an object to something that doesn't execute? I don't think it is. It seems like it would be really tricky to check to make sure every exec flag asked for by every object actually gets run, and the current system doesn't guarantee that either.

**Comment by aeslaughter** [comment]
The current system does, just programmatically. No object has an execute flag listed that would not get called (to my knowledge at least), with the exception of EXEC_CUSTOM, which has always been that way.

**Comment by friedmud** [comment]
I don't really see the problem here - just check that the app has
registered each execute_on flag in checkProblemIntegrity

**Comment by permcody** [comment]
I agree with Derek, no need to try to figure out if there's an integrity
problem in each object. Just figure out all the options and valid
combinations in the integrity check.
